client/state/selectors/get-countries.ts

```diff
@@ -0,0 +1,16 @@
+import type { CountriesState } from '../countries/types';
+import type { IAppState } from '../types';
+
+import 'calypso/state/countries/init';
+
+/**
+ * Retrieves the list of countries from the specified type.
+ *
+ * The country data types are 'domains, 'payments', or 'sms'.
+ */
+export default function getCountries< TType extends keyof CountriesState >(
+	state: IAppState,
+	type: TType
+) {
+	return state.countries?.[ type ] ?? null;
```

Comment:
Since we're including the reducer (L4), `state.countries` is expected to always be there, therefore `state.countries` shouldn't need optional chaining, right?

Reply:
I did this because in the non-TS code the line was `get( state, [ 'countries', type ], null )` which corresponds to `state.countries?.[ type ] ?? null` and I didn't want to change the behavior. 

On the other hand, if the types are right and the `init` file has correctly registered the reducer, you're right that `countries` will always exist.

On the other other hand, unit tests in calypso often create a Redux store without the code that allows reducers to register themselves and leave out parts of state that they do not need. I wouldn't be surprised if removing this optional chaining could break some tests. I'll give it a try and find out.

Reply:
I pushed an update to do option 3, so this should be resolved now.

Reply:
From 3), how is it possible that the `countries` reducer will not be registered? With the modular state, we've been making sure that every selector and action that make use of that state tree will include that reducer, so there should never be the scenario where that tree doesn't exist.

Furthermore, I do agree that in tests it can be difficult to get it all working with requiring all required reducer trees and their corresponding initial state. To fix that, we do need a helper and I appreciate you working on that. One idea for it could be including the `addReducerEnhancer` enhancer when creating the mock store. That will allow modular reducers to dynamically get included in the state tree as relevant selectors and actions initialize those reducers. Then the only thing we might need to provide would be initial state when necessary. 

This does go a bit beyond what the purpose of this PR is, but let me know what you think.